libsolidity/analysis/FunctionCallGraph.h

```diff
@@ -0,0 +1,98 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+
+#include <vector>
+#include <variant>
+
+namespace solidity::frontend
+{
+
+/**
+ * Creates a Function call graph for a contract at the granularity of Solidity
+ * functions and modifiers
+ *
+ * Includes the following special nodes:
+ *  - CreationRoot: All calls made at contract creation originate from this node
+ *  - CreationDispatch: Represents the internal dispatch function at creation time
+ *  - ExternalDispatch: Represents the runtime dispatch for all external functions
+ *  - InternalDispatch: Represents the runtime dispatch for internal function pointers and complex expressions
+ *
+ *  Nodes are a variant of either the enum SpecialNode or an ASTNode pointer.
+ *  ASTNodes are usually inherited from CallableDeclarations
+ *  (FunctionDefinition, ModifierDefinition, EventDefinition) but for functions
+ *  without declaration it is directly the FunctionCall AST node.
+ *
+ *  Functions that are not called right away as well as functions without
+ *  declarations have an edge to the internal dispatch node.
+ *
+ *  Auto-generated getter functions for public state variables are ignored.
+ */
+class FunctionCallGraphBuilder: private ASTConstVisitor
+{
+public:
+	enum class SpecialNode { CreationRoot, CreationDispatch, InternalDispatch, ExternalDispatch };
```

Comment:
```suggestion
	enum class SpecialNode { CreationRoot, InternalDispatchCreation, InternalDispatch, ExternalDispatch };
```
?
I think the two internal dispatch nodes should have more in common...

`ExternalDispatch` and `CreationRoot` are also essentially the same thing just for creation / deployed.

Reply:
Maybe
 - EntryCreation
 - InternalDispatchCreation
 - Entry
 - InternalDispatch
?

Reply:
> ExternalDispatch and CreationRoot are also essentially the same thing just for creation / deployed.

Really? I mean, all the initializations for state variables are linked to that as well as function calls in the parameters of the class constructors and such... I am not sure that has anything to do with dispatch at all...

Reply:
The external dispatch is the entry point after deployment. The creation root is the entry point at creation time.

Reply:
Looks like this has already been fixed.